lib/datadog/appsec/utils/read_write_lock.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Datadog
+  module AppSec
+    module Utils
+      # Simple implentation for a fair read write lock
+      # It allows for as many concurrent readers but only one concurrent writer.
+      # To avoid the writer thread being exhausted, once a writer thread has signaled that it wants to acquire the lock,
+      # no more reader threads can acquire the lock.
+      # The new reader threads will wait for the writer thread to finish.
+      class ReadWriteLock
+        def initialize
+          @reader_mutex    = Mutex.new
+          @reader_q        = ConditionVariable.new
+          @reader_count    = 0
+          @reader_releases = 0
+
+          @writer_mutex    = Mutex.new
+          @writer_q        = ConditionVariable.new
+          @writter         = false
+        end
+
+        def with_rlock
+          rlock
+          yield
+        ensure
+          runlock
+        end
+
+        def with_lock
+          lock
+          yield
+        ensure
+          unlock
+        end
+
+        def rlock
+          @reader_mutex.synchronize do
+            @reader_count += 1
+
+            writter = false
+
+            @writer_mutex.synchronize do
+              writter = @writter
+            end
+
+            @reader_q.wait(@reader_mutex) if writter
+          end
+        end
+
+        def runlock
+          @reader_mutex.synchronize do
+            @reader_releases += 1
+
+            @writer_mutex.synchronize { @writer_q.signal } if @reader_releases == @reader_count
+          end
+        end
+
+        def lock
```

Comment:
Consider checking if the writer lock is already taken -- right now if a writer tries to acquire a lock twice, it will either succeed (if there are no readers), and then incorrectly release the lock after the first unlock (oops), or it will deadlock (if there are active readers waiting -- I leave why that is as an exercise for the reader).